### PR TITLE
fix: save last finalized block

### LINKED ISSUE #1712: Save last finalized block
### Describe the feature

To be more accurate on what last finalized block is, we should save this information inside database, this would help a little bit on initial sync to BlockchainTree and SyncController to be more concise on what blocks need to be reorged and what blocks are finalized.

Currently we are just taking `tip-finalization_window`, that still works.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by fgimenez**
Closes #1712

Prevents potential issues with fixed finality depth on adverse conditions, there are no fixed bounds to finality depth and the actual finalized block can be earlier than a  `last_finalized_block_number` initialized to a fixed value.

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
lgtm, storing finalized is sound imo because finalized, hehe

pending @rkrasiuk

**Review by onbjerg — APPROVED** [review]
makes sense to me

**Comment by mattsse** [comment]
> unwinds should probably clear this one out if it goes beyond this block. this can happen when manually calling unwind

what are you suggesting? what needs to be done?

**Comment by fgimenez** [comment]
> if there's a pipeline unwind that goes beyond that block, replace it with a lower one on Finish stage unwind fn

@joshieDo cool thx! we should save the lowest one that results from the unwind operation, right? either `to-block` or the lowest end resulting from `num-blocks`?

**Comment by fgimenez** [comment]
> if there's a pipeline unwind that goes beyond that block, replace it with a lower one on Finish stage unwind fn
@joshieDo ok done here https://github.com/paradigmxyz/reth/pull/8473/commits/eaec49db6cb64b779232c05d4b5321cb99bb884f by introducing a `FinalizedBlockProvider` trait with default methods, ptal

File: crates/storage/provider/src/traits/finalized_block.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        let last_finalized_block_number = finalized_blocks.pop_first().unwrap_or_default();
+        Ok(last_finalized_block_number.1)
+    }
+
+    /// Saves finalized block.
+    fn save_finalized_block_number(&self, block_number: BlockNumber) -> ProviderResult<()> {
+        let provider = self.provider_factory().provider_rw()?;
+        provider.tx_ref().put::<tables::FinalizedBlocks>(FINALIZED_BLOCKS_KEY, block_number)?;
+        provider.commit()?;
```
[the remark this answers is not in the document]
Comment (fgimenez) [reply crates/storage/provider/src/traits/finalized_block.rs]: indeed, thx for the pointer. I've split the finalized block trait into a writer an a reader and removed the default implementations, they are now part of the impl for `DatabaseProvider<TX>`, similar to the rest of trait implementations, ptal

File: bin/reth/src/commands/stage/unwind.rs [hunk 2]
```diff
@@ -119,6 +141,15 @@ impl Command {
...
 
+        // update finalized block if needed
+        let finalized_block_manager = FinalizedBlockManager::new(provider_factory.clone());
+        let last_finalized_block = finalized_block_manager.fetch_latest_finalized_block_number()?;
+        let range_min =
+            range.clone().min().ok_or(eyre::eyre!("Coul not fetch lower end of range"))?;
+        if range_min < last_finalized_block {
+            finalized_block_manager.save_finalized_block_number(BlockNumber::from(range_min))?;
+        }
+
```
[the remark this answers is not in the document]
Comment (fgimenez) [reply bin/reth/src/commands/stage/unwind.rs]: I've moved it to the unwind pipeline stages loop, before we do the `provider_rw.commit`, ptal.

File: bin/reth/src/commands/stage/unwind.rs [hunk 3]
```diff
@@ -68,6 +68,28 @@ pub struct Command {
     command: Subcommands,
 }
 
+struct FinalizedBlockManager<DB> {
```
[the remark this answers is not in the document]
Comment (fgimenez) [reply bin/reth/src/commands/stage/unwind.rs]: nope, without the default trait impl there's no need for this :+1:

**Review by joshieDo — APPROVED** [review]

File: crates/storage/provider/src/providers/database/provider.rs [hunk 4]
```diff
@@ -2753,6 +2753,29 @@ impl<TX: DbTx> StatsReader for DatabaseProvider<TX> {
     }
 }
 
+/// Table key used to read and write the last finalized block.
+const FINALIZED_BLOCKS_KEY: u8 = 0;
+
+impl<TX: DbTx> FinalizedBlockReader for DatabaseProvider<TX> {
+    fn fetch_latest_finalized_block_number(&self) -> ProviderResult<BlockNumber> {
```
Comment (rkrasiuk) [inline crates/storage/provider/src/providers/database/provider.rs:2760]:
```suggestion
    fn last_finalized_block_number(&self) -> ProviderResult<BlockNumber> {
```

File: crates/storage/db/src/tables/mod.rs [hunk 5]
```diff
@@ -406,6 +406,9 @@ tables! {
 
     /// Stores EIP-7685 EL -> CL requests, indexed by block number.
     table BlockRequests<Key = BlockNumber, Value = Requests>;
+
+    /// Stores the last finalized block.
+    table FinalizedBlocks<Key = u8, Value = BlockNumber>;
```
Comment (rkrasiuk) [inline crates/storage/db/src/tables/mod.rs:411]: if we name the table this way, we won't be able to reuse it for anything else. pls consider giving some generic name

Comment (fgimenez) [reply crates/storage/db/src/tables/mod.rs]: makes sense, renamed to `ChainState`, ptal

**Pull request merged**